daemon/logger/jsonfilelog/read.go

```diff
@@ -121,96 +119,106 @@ func tailFile(f io.ReadSeeker, logWatcher *logger.LogWatcher, tail int, since ti
 	}
 }
 
-func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
-	dec := json.NewDecoder(f)
-	l := &jsonlog.JSONLog{}
+func (l *JSONFileLogger) followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
+	var (
+		msg        *logger.Message
+		err        error
+		rotated    bool
+		readIndex  uint64
+		writeIndex uint64
+
+		dec        = json.NewDecoder(f)
+		log        = &jsonlog.JSONLog{}
+		subscriber = l.writeNotifier.Subscribe()
+		watchClose = logWatcher.WatchClose()
+	)
+
+	go func() {
+		for {
+			if _, ok := <-subscriber; !ok {
```

Comment:
Since there is no timeout, the publisher should not block at all.
We only ever need 1 write notification at a time, and since we have a channel buffer of 1 it will be able to store up the write notification for us until we're ready for it.